[![DOI](https://zenodo.org/badge/1150695954.svg)](https://zenodo.org/badge/latestdoi/1150695954)

# Repo for the analysis of Fks hotspots DMS

Welcome to the new-and-improved companion repo for [this manuscript](https://www.biorxiv.org/content/10.1101/2024.07.21.604487).

It is essentially divided in two sections: `pre/`, which corresponds to preliminary steps and analyses and `post/`, which corresponds to downstream analysis of selection coefficients obtained using gyōza (as well as analysis of experiments performed after the DMS).

## Preliminary steps

### Homologous hotspot sequences

A multiple sequence alignment was built from Fks homologous sequences retrieved using [MetaPhOrs](https://orthology.phylomedb.org/). We first kept ["Fks1" hits](./pre/ortholog_data/FKS1_filtered.aln) separate from ["Fks2" hits](./pre/ortholog_data/FKS2_filtered.aln) in order to extract the hotspot sequences which were used to generate the oPools. Taxonomic data were obtained by using's [NCBI's CommonTree](https://www.ncbi.nlm.nih.gov/Taxonomy/CommonTree/wwwcmt.cgi) with [the full list of strains](./pre/ortholog_data/allspecies.txt) for which there was at least one Fks hit. [This custom R script](./pre/scripts/generate_S3_data.R) was then used to generate [Supplementary Table 3](./pre/ortholog_data/S3_data.csv). Ultimately all hits were pooled to generate a single multiple sequence alignment from which we generated the phylogenetic tree in Supplementary Figure 3A.

[This short script](./pre/scripts/EL4.py) was used to measure conservation of the EL4 flexible loop Fks homologs.

### Generate oPools

oPools were generated using [this notebook](./pre/notebooks/generate_oPools.ipynb).

### Preliminary growth assays

Deletion mutants for the hotspot 1 of FKS1 and FKS2 + the parental strain were cultured in increasing drug concentrations to generate dose-reponse curves, using [this notebook](./pre/notebooks/dose_response.ipynb).

### DMS cultures

[This notebook](./pre/notebooks/dms_cultures.ipynb) was used to visualize the number of mitotic generations (calculated from cell concentrations measured by flow cytometry) and look at the correlation between cytometry measurements and optical density measurements. **Important**: a single dataframe was used to store all information on DMS samples. In other words, the output file generated by the notebook was used to get the sample layout required by gyōza and fill in the number of mitotic generations for each condition.

## gyōza

The first step was to analyze the DMS dataset with [gyōza v1.2.1](https://github.com/durr1602/gyoza)

Single mutants of Hotspots 1, 2 and 3 were analyzed using [this config](./config/config.yaml).

For simplicity, Fks homologous hotspots were analyzed **separately** using the `provided` mode of gyōza. This mode required that expected sequences be provided in a specific format. To generate the proper input (silent mutants from the first dataset + orthologous hotspot sequences), we ran [this custom script](./pre/scripts/generate_gyoza_input_orthologs.py).

For all gyōza analyses, read count threshold at T0 was set at `10`, which means only variants (amino acid sequences) which had more than `10` reads in **all** T0 replicates were used to calculate an average selection coefficient across replicates.

Similarly, sequencing data from preliminary MiSeq/Aviti runs were analyzed with gyōza by disabling the `process_frequencies` config key. This allowed estimating library diversity, as well as estimating bottlenecks due to deleterious effects during the overnight culture.

## Requirements

The following steps need to be run in a Python 3.13 virtual environment. With `uv`:

```
uv venv fks_gyoza --python 3.13
source fks_gyoza/bin/activate
uv pip install -r post/requirements.txt
```

## Analysis

### How to reproduce analyses
Simply run [this notebook](./post/notebooks/driver.ipynb)

### Estimation of library bottlenecks
Allele frequencies for libraries of single mutants and homologous hotspot sequences were compared before/after overnight culture. (before = MiSeq/Aviti preliminary sequencing runs to estimate diversity in libraries, after = T0 of main sequencing runs). The corresponding dataframes (obtained from different gyōza runs) were manually copied to two single locations (`pre/freq_libraries` and `results/df/distribution_freq`) before being compared using [this notebook](./post/notebooks/bottlenecks.ipynb).

### Estimation of replicability
Spearman correlation coefficients were calculated for each pairwise comparison of replicates using [this script](./post/scripts/plot_replicates.py). We also compared results for FKS1-HS1 and FKS1-HS2 with or without repression of *FKS2* using [this script](./post/scripts/BYvR11.py).

### Classification of scores to get labels
A Gaussian Mixture Model (GMM) was [trained](./post/notebooks/train_GMM.ipynb) and used to [classify fitness scores](./post/notebooks/classify_gyoza_data.ipynb) calculated by gyōza to obtain labels reflecting mutational effects (deleterious, WT-like, intermediary, resistant, etc).

Thresholds were set to resolve overlaps between the different gaussians predicted.

Classification was collapsed (deleterious/non-deleterious in the control condition, sensitive/resistant upon selection).

Classified data were plotted using [this notebook](./post/notebooks/plot_classified_data.ipynb)

### Comparison with FungAMR
[The FungAMR database](https://card.mcmaster.ca/fungamrhome) was downloaded on April 7th, 2025, then parsed using [this script](./pre/scripts/Fks1_ortholog_matuted_seq_fungAMR.py) to convert mutations previously reported in the literature to *S. cerevisiae* numbering, generating [a raw version of Supplementary Table 1](./pre/fungamr/Fks_mutated_hotspot_seq_160126.csv).

[This notebook](./post/notebooks/compare_to_fungamr.ipynb) was then used to compare what has been reported (mostly MICs) to our classified data, generating a fragmented version of Supplementary Table 2 (e.g. [FKS1-HS1](./post/fungamr/agreements_BY4741_FKS1-HS1.csv)).

### Validations with reconstructed mutants
DMS data were [compared to growth data](./post/notebooks/20240129_validations_test3.ipynb) from reconstructed mutants to perform a linear regression and infer a score for a few single mutants in FKS1-HS1 missing from the DMS dataset. Note: the mutants to be reconstructed were selected because they spanned the full fitness range. This was done a while ago using a notebook that I won't include in this repo but is accessible [here](https://github.com/Landrylab/Durand_et_al_2024/blob/fefd36630aa73609d9b551f08b29505e646d8769/select_mutants_for_reconstruction.ipynb).

A second validation experiment was performed with more mutants, this time outside the hotspots. The corresponding growth curves were analyzed and transformed using [this notebook](./post/notebooks/20250210_more_mutants.ipynb).

### Plot heatmaps
[This notebook](./post/notebooks/heatmaps.ipynb) was used to plot heatmaps of fitness scores.

### Generate defattr files for ChimeraX
[This notebook](./post/notebooks/defattr.ipynb) was used to generate the files used in ChimeraX to plot onto the protein structures the proportion of resistant mutants per position for each hotspot.

### Generate logos from homologous hotspot sequences
[This notebook](./post/notebooks/orthologs.ipynb) was used to find associations between residues at each position and the resistance phenotype of homologous hotspot sequences spiked in the DMS experiment.

## How the figures and tables were generated

Main figures:
- Figure 1: MD simulations, figures generated in PyMOL
- Figure 2A: Structural alignment, figure generated in ChimeraX
- Figure 2B: Experimental design, figure created with BioRender
- Figure 2C-H: Distributions of mutational effects and UpsetPlots of classified data, figures generated with [this template notebook](./post/notebooks/plot_classified_data.ipynb)
- Figure 3A-L: Heatmaps of mutational effects for Fks1, figures generated with [this template notebook](./post/notebooks/heatmaps.ipynb)
- Figure 3M-P: Mutational effects mapped onto the structure of Fks1 using attribute files generated with [this template notebook](./post/notebooks/defattr.ipynb) and ChimeraX

Supplementary figures:
- Supplementary Figure 1: MD simulations, figures generated in PyMOL
- Supplementary Figure 2: MD validations with growth data from reconstructed mutants analyzed with [this notebook](./post/notebooks/20250210_more_mutants.ipynb)
- Supplementary Figure 3A: Phylogenetic tree (see [corresponding section](#homologous-hotspot-sequences))
- Supplementary Figure 3B: Structural alignment, figure generated in ChimeraX
- Supplementary Figure 4: DMS growth data analyzed with [this notebook](./pre/notebooks/dms_cultures.ipynb)
- Supplementary Figure 5: Allele frequencies at T0, plotted with [this notebook](./post/notebooks/bottlenecks.ipynb)
- Supplementary Figure 6: Replicability, plotted with [this script](./post/scripts/plot_replicates.py)
- Supplementary Figure 7: Gaussian mixture model trained with [this template notebook](./post/notebooks/train_GMM.ipynb)
- Supplementary Figure 8: DMS validations with growth data from reconstructed mutants analyzed with [this notebook](./post/notebooks/20240129_validations_test3.ipynb)
- Supplementary Figure 9A-J: Heatmaps of mutational effects for Fks2, figures generated with [this template notebook](./post/notebooks/heatmaps.ipynb)
- Supplementary Figure 9K-O: Mutational effects mapped onto the structure of Fks1 using attribute files generated with [this template notebook](./post/notebooks/defattr.ipynb) and ChimeraX
- Supplementary Figure 10: Effect of paralog repression, plotted with [this script](./post/scripts/BYvR11.py)
- Supplementary Figure 11: Logos of homologous hotspot sequences, generated with [this notebook](./post/notebooks/orthologs.ipynb)
- Supplementary Figure 12: Dose-response curves generated with [this notebook](./pre/notebooks/dose_response.ipynb)

Supplementary tables:
- Supplementary Tables 1 and 2: FungAMR data (see [corresponding section](#comparison-with-fungamr))
- Supplementary Table 3: List of homologous hotspots generated using [this R script](./pre/scripts/generate_S3_data.R)
- Supplementary Table 4: Table of strains, plasmids and primers, mainly generated by hand. Also includes oPools generated with [this notebook](./pre/notebooks/generate_oPools.ipynb)

Final figures were assembled and finalized in Inkscape.